observatory/tracking.py

```diff
@@ -1,51 +1,247 @@
+            raise AssertionError(
+                f'Could not find source file {absolute_file_path} ' +
+                'to upload as output of this run. Please make ' +
+                'sure that the file exists on disk.')
+
+        self._state.record_output(
+            self.name, self.version, self.experiment,
+            self.run_id, filename, open(absolute_file_path))
+
+    def __enter__(self):
+        self._state.record_session_start(
+            self.name, self.version, self.experiment, self.run_id)
+
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        if exc_type is None:
+            session_status = 'COMPLETED'
+        else:
+            session_status = 'FAILED'
+
+        self._state.record_session_end(
+            self.name, self.version, self.experiment,
+            self.run_id, session_status)
+
+
+        return exc_type is None
+
+class ObservatoryState(ABC):
+    def __init__(self):
+        self.n = None
+
+    def switch(self, state):
+        self.__class__ = state
+
+    @abstractmethod
+    def record_metric(self, name, value):
+        """
+        This method will record a metric.
+        """
+        pass
+
+    @abstractmethod
+    def record_settings(self, name, value):
+        """
+        This method will record a setting.
+        """
+        pass
+
+    @abstractmethod
+    def record_output(self, name, value):
+        """
+        This method will record a output.
+        """
+        pass
+
+    @abstractmethod
     def record_session_start(self, model, version, experiment, run_id):
         """
-        Records the start of a session.
+        This method will record the start of a run.
+        """
+        pass
+
+    @abstractmethod
+    def record_session_end(self, model, version, experiment, run_id, status):
+        """
+        This method will record the end of a run.
+        """
+        pass
+
+class LocalState(ObservatoryState):
+    #Handels metrics for the local filesystem -> Sink.py
+
+    def record_metric(self, model, version, experiment, run_id, name, value):
+        #sink.save_metric(model, version, experiment, run_id, name, value)
+        #localstate is nothing more than a nice data handler that passes is to sink.py
+        #this is because the sever is also going to use sink.py to save data
+        print("LocalState : record_metric")
+
+    def record_settings(self, model, version, experiment, run_id, settings):
+        print("LocalState : record_settings")
+
+    def record_output(self, model, version, experiment, run_id, filename, file):
+        print("LocalState : record_output")
+
+    def record_session_start(self, model, version, experiment, run_id):
+        print("LocalState : record_session_start")
+
+    def record_session_end(self, model, version, experiment, run_id, status):
+        print("LocalState : record_session_end")
+
+
+class RemoteState(ObservatoryState):
+    #Handels metrics for the remote endpoint
```

Comment:
Can you convert this to a proper documentation string?

```suggestion
"""
Records metric on a remote server that you can run through the command `observatory server`.
"""
```